### PR TITLE
feat: add @typescript-eslint/recommended-type-checked

### PR DESCRIPTION
**Pull request by blacha**
#### Motivation

Adds more complex type checking including rules like requiring await for promises and async functions.

full list of rules can be found https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/configs/recommended-type-checked.ts

#### Modification

adds typescript-eslints recommended type checking rules

an additional `project: true` was needed so that eslint can use the type checker from typescript

### CODE REVIEW

File: src/__test__/class.test.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@ export class FooBar {
   }
 
   async bar(): Promise<string> {
-    return 'bar';
+    return await Promise.resolve('bar');
```
Comment (jfee-linz) [inline src/__test__/class.test.ts:7]: I think the await is optional here (just `return Promise.resolve...` unless the value is used within the method)

**Review by jfee-linz — APPROVED** [review]

File: src/__test__/class.test.ts (hunk 1 above)
Comment (blacha) [reply src/__test__/class.test.ts:7]: the await is not optional as the method is `async` there is no point in the method being `async` unless a `await` is used.

granted this is a very useless method :)

Comment (jfee-linz) [reply src/__test__/class.test.ts:7]: my understanding is async only requires a promise to be returned, but it usually only makes sense to await when the value is used within a method.

**Review by philals — APPROVED** [review]

File: src/__test__/class.test.ts (hunk 1 above)
Comment (blacha) [reply src/__test__/class.test.ts:7]: Its quite nitpicking but, async functions always return a promise, but they only need to be `async` if `await` is used.

I believe their is a performance penalty for using them so in the below example the second function is more correct. :shrug: 

```typescript
async foo() { return "bar" } 

foo() // Promise<"bar">

// More correct ?
fooProm() { return Promise.resolve("bar") }

fooProm() // Promise<"bar">

```

Comment (dalmolinz) [reply src/__test__/class.test.ts:7]: I think `await Promise.resolve` is redundant. You could either `return await 'bar'`, or `return Promise.resolve('bar')` (and in this case not use `async` either). The consumer would need to await/.then it anyway.

**Pull request merged**